src/jit/optimizer.cpp

```diff
@@ -7045,33 +7066,39 @@ void Compiler::optComputeLoopSideEffectsOfBlock(BasicBlock* blk)
                                 // and might have arbitrary side effects.
                                 if ((tree->gtFlags & GTF_CALL_HOISTABLE) == 0)
                                 {
-                                    heapHavoc = true;
+                                    memoryHavoc |= memoryKindSet(GcHeap, ByrefExposed);
                                 }
                             }
                         }
                         else
                         {
-                            heapHavoc = true;
+                            memoryHavoc |= memoryKindSet(GcHeap, ByrefExposed);
                         }
                         break;
                     }
 
                     default:
-                        // All other gtOper node kinds, leave 'heapHavoc' unchanged (i.e. false)
+                        // All other gtOper node kinds, leave 'memoryHavoc' unchanged (i.e. false)
                         break;
                 }
             }
         }
     }
 
-    if (heapHavoc)
+    if (memoryHavoc != 0)
```

Comment:
Since we assume all object field stores, static field stores, and array element stores alias all byref loads, we [treat them](https://github.com/dotnet/coreclr/pull/9169/commits/c0b13389b3cb5fe9d16127ae0eebf135589b8760#diff-15cd21de3f1ffaa677e7da145576d60cR6952) as `ByrefExposed`-havoc (but not `GcHeap`-havoc).  So every loop that has these but not `GcHeap`-havoc will have one kind of havoc but not the other.